File: prolog/old/opts.pl
:- writeln("-- opts.pl --").
:- ensure_loaded(["gb_env.pl","funcs.pl","tstamp.pl","dhash.pl","status.pl"]).

:- dynamic pass/1.

:- retractall(pass/1).
:- dynamic gt/2.

incp(P1) :-
    (   pass(P1)
    ->  retract(pass(P1)),
        P2 is P1+1,
        assert(pass(P2))
    ;   P1 is 1,
        assert(pass(P1))
    ).

clean :-
    \+ dirty1(_, _, _, _, _, _).

nodeps(T) :-
	writeln('-nodeps'),
    \+ dep(_, T).

dag(G1) :-
	writeln('-dag'),
    findall(Sa-Ta, dep(Ta, Sa), Egs),
    add_edges([], Egs, G1).

nts(G) :-
	writeln('-nts'),
    dag(G1),
    top_sort(G1, G).

dagt(G2) :-
	writeln('-dagt'),
    dag(G1),
    transpose_ugraph(G1, G2).

gproc3((T, Opl, Res)) :-
	write('-gproc3 '),
	writeln(T),
    Op=..Opl,
    (   call(Op)
    ->  Res=true,
        incp(P1),
        updatets(P1)
    ;   Res=false
    ).

gproc2(T) :-
	writeln('-gproc2'),
    findall([Op, Tl, Si, Dd, Sd, Opt],
            (	dirty1(Op, Tl, Si, Dd, Sd, Opt),
            	member(T, Tl)
            ),
            L1),
    (   L1==[]
    ->  true
    ;   sort(L1, L2),
        findall((T, X, _), member(X, L2), L3),
        maplist(gproc3, L3),
        \+ member((T, _, false), L3)
    ).

gproc([]) :-
    writeln('-gproc[]').
gproc([T|Tl]) :-
	writeln('-gproc'),
    gproc2(T),
    gproc(Tl).

opexec :-
	writeln('-opexec'),
    retractall(pass/1),
    nts(G1),
    !,
    writeln(G1),
    incp(P1),
    updatets(P1),
    (   gproc(G1)
    ->  (   clean
        ->  halt(0)
        ;   fail
        )
    ;   halt(1)
    ).

File: devbak/opts.py
from bkenv import *
from status import *
from toposort import topological_sort
import asyncio

_pass = 1


def incp():
    global _pass
    i = _pass
    _pass += 1
    return i


def clean():
    res = len(list(dirty1())) == 0
    if res:
        print('clean')
    return res


def nodeps(T):
    # print('-nodeps')
    for (T1, S1) in dep:
        if S1 == T:
            return False
    return True


def istgt(T):
    # print('-istgt')
    for (T1, S1) in dep:
        if T1 == T:
            return True
    return False


def nts():
    print('-nts')
    p1 = topological_sort(dep)
    ts = [t for elem in p1 for t in elem]
    return [d for d in ts if istgt(d)]


running = set()


async def gproc3(L):
    print('-gproc3-', end='')
    if L[1] not in running:
        print(L[0], L[1])
        running.add(L[1])
        if nodeps(L[0]):
            async def f1(L):
                try:
                    L[2] = await L[1]()
                    running.remove(L[1])
                except Exception as e:
                    print(e)
            tsk = asyncio.create_task(f1(L))
        else:
            try:
                L[2] = await L[1]()
                running.remove(L[1])
            except Exception as e:
                print(e)
        
    # else:
    #print(Op, 'already running')


async def gproc2(T):
    # print('-gproc2')
    # print(T)
    L1 = list(dirty1(T))
    if len(L1) == 0:
        return True
    L2 = [[T, Op, (0, 0)] for Op in L1]
    for li in L2:
        await gproc3(li)
    L3 = [R for (T, Op, R) in L2 if R[1]]
    return len(L3) == 0


async def gproc(L):
    # print('-gproc')
    for T in L:
        if not await gproc2(T):
            return False
    return True


async def opexec():
    print('opexec')
    ic = 0
    G1 = nts()
    print(G1)
    p1 = incp()
    # updatets(p1)
    while True:
        if await gproc(G1):
            if clean():
                return True
            else:
                return False
        else:
            return False


if __name__ == '__main__':
    updatets(0)
    print(asyncio.run(opexec()))
